modules/docs/src/05-design/03-services.md

```diff
@@ -0,0 +1,293 @@
+
+Finally-encoded KVStore algebra :
+
+```scala
+trait KVStore[Context[_]]{
+  def put(key: String, value: String) : Context[Unit]
+  def get(key: String)                : Context[Option[String]]
+  def delete(key: String)             : Context[Unit]
+}
+```
+
+Initially-encoded KVStore algebra :
+
+```scala
+sealed trait KVStoreOp[Output]
+object KVStoreOp {
+  case class Put(key: String, value: String)  extends KVStore[Unit]
+  case class Get(key: String)                 extends KVStore[Option[String]]
+  case class Delete(key: String)              extends KVStore[Unit]
+}
+```
+
+These two encodings contain a similar amount of information. It is even nearly-trivial to go from a `KVstore[Context]` instance to a `KVStoreOp ~> Context` natural transformation, and vice versa:
+
+```scala
+trait ~>[F[_], G[_]]{
+  def apply[A](fa: F[A]) : G[A]
+}
+
+def asNaturalTransformation[Context[_]](impl: KVStore[Context]) = new (KVStoreOp ~> Context){
+  def apply[A](fa: KVStoreOp[A]) : Context[A] = fa match {
+    case KVStoreOp.Put(key, value) => impl.put(key, value)
+    case KVStoreOp.Get(key)        => impl.get(key)
+    case KVStoreOp.Delete(key)     => impl.delete(keyu)
+  }
+}
+
+def fromNaturalTransformation[Context[_]](run: KVStoreOp ~> Context) = new KVStore[Context]{
+  def put(key: String, value: String) = run(KVStoreOp.Put(key, value))
+  def get(key: String)                = run(KVStoreOp.Get(key))
+  def delete(key: String)             = run(KVStoreOp.Delete(key))
+}
+```
+
+This duality is heavily used by Smithy4s : finally-encoded interfaces are generally more natural to Scala developers, and are better supported in editors (autocompletion, etc). But from an implementation's perspective, the initial, data-base encoding is really interesting, because operations are reified as **datatypes** that can be associated with instances of generic typeclasses : it is possible to abstract over data, it is not possible to abstract over method calls.
+
+### A detour around kinds
+
+The methods generated by Smithy4s are conceptually similar to the methods expressed in the example above, except that the output types are significantly more verbose.
+
+```scala
+trait Interface[Context[_, _, _, _, _,]]{
+  def operation1(a: A, b: B) : Context[Input, Error, Output, StreamedInput, StreamedOutput]
+}
+```
+
+Let's address this awkwardness right away, by explaining the rationale behind this seemingly humongous signature :
+
+#### Input
+
+That's the input type of an operation. Typically, a case class that holds fields matching the method parameters. We keep track of it in the output type for several reasons:
```

Comment:
```suggestion
That's the input type of an operation. Typically, a case class that holds fields matching the method parameters. We keep track of it in the return type for several reasons:
```

Wasn't sure if you meant output of the function or the `Output` parameter.. if that even makes sense 😆